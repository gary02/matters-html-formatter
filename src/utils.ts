import * as cheerio from "cheerio"

export const stripHtml = (html: string, replacement = " ") =>
  (String(html) || "")
    .replace(/(<\/p><p>|&nbsp;)/g, " ") // replace line break and space first
    .replace(/(<([^>]+)>)/gi, replacement)

export const makeSummary = (html: string, length = 140) => {
  // buffer for search
  const buffer = 20

  // split on sentence breaks
  const sections = stripHtml(html, "")
    .replace(/([?!。？！]|(\.\s))\s*/g, "$1|")
    .split("|")

  // grow summary within buffer
  let summary = ""
  while (summary.length < length - buffer && sections.length > 0) {
    const el = sections.shift() || ""

    const addition =
      el.length + summary.length > length + buffer
        ? `${el.substring(0, length - summary.length)}...`
        : el

    summary = summary.concat(addition)
  }

  return summary
}

/**
 * Output clean HTML for IPFS
 */
export const outputCleanHTML = (html: string) => {
  /**
   * Note: enable `xmlMode` to remove default wrapper
   *
   * @see https://github.com/cheeriojs/cheerio/issues/1031#issuecomment-368307598
   */
  const $ = cheerio.load(html, { decodeEntities: false, xmlMode: true })

  // remove audio player
  $(".player").remove()

  return $.html()
}
