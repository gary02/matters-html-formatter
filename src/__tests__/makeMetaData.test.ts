import { makeMetaData } from '..'
import { testMetaData } from './utils'

describe('makeMetaData', () => {
  test('can generate content meta data', () => {
    expect(makeMetaData(testMetaData)).toMatchSnapshot()
  })
})
