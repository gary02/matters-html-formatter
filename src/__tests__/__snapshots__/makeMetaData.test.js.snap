// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeMetaData can generate content meta data 1`] = `
Object {
  "@context": "http://schema.org",
  "@id": "ipfs://ipfs-hash",
  "@type": "Article",
  "author": Object {
    "description": "this is a test user",
    "name": "Test User",
    "url": "user-home-page",
  },
  "dateCreated": "2020-12-23",
  "description": "This is a piece of test content",
  "image": "image-url",
}
`;
